File: Arrays and Strings/1.4-Palindrome-Permutation/PalPerm.java
import java.util.*;

public class PalPerm {

    public boolean isPalPerm(String str) {
	String lstr = str.toLowerCase();
	HashMap<Character, Integer> counts = new HashMap<Character, Integer>();
	int numChars = 0;

	for (int i = 0; i < lstr.length(); ++i) {
	    char c = lstr.charAt(i);
	    if (c != ' ')  {
		/* This is why I hate you Java */
		if (!counts.containsKey(c)) {
		    counts.put(c, 1);
		} else {
		    counts.put(c, counts.get(c) + 1);
		}
		++numChars;
	    }
	}

	boolean evenLength = numChars % 2 == 0;
	boolean oneOddCount = false;

	for (Map.Entry<Character, Integer> entry : counts.entrySet()) {
	    if (entry.getValue() % 2 != 0) {
		if (evenLength) return false;
	        else if (oneOddCount) return false;
		else oneOddCount = true;
	    }
	}

	return true;
    }

    public static void main(String[] args)  {
	System.out.println("Enter a string: ");
	Scanner in = new Scanner(System.in);

	PalPerm pal = new PalPerm();
	System.out.println("The string " + (pal.isPalPerm(in.nextLine()) ? "is" : "isn't") + " a permutation of a palindrome.");

	in.close();
    }
}

File: 01 Arrays and Strings/Arrays and Strings/1.4-Palindrome-Permutation/palperm.cpp
#include <iostream>
#include <string>
#include <cctype>
#include <unordered_map>

bool isPalPerm(std::string str) {
  std::unordered_map<char, int> counts;
  int numChars = 0;

  for (char c : str) {
    if (c != ' ') {
      ++numChars;
      ++counts[std::tolower(c)];
    }
  }

  bool evenLength = numChars % 2 == 0;
  bool oneOddCount = false;

  for (std::pair<const char, int>& entry : counts) {
    if (entry.second % 2) {
      /* If there are an even number of characters and a character appears an
	 odd number of times */
      if (evenLength) return false;
      /* Odd number of characters and more than one character appears an odd
	 number of times */
      else if (oneOddCount) return false;
      /* First character to appear an odd number of times */
      else oneOddCount = true;
    }
  }

  return true;
}

int main() {
  std::cout << "Enter a string: " << std:: endl;
  std::string str;
  std::getline(std::cin, str);
  bool palperm = isPalPerm(str);
  std::cout << "The string " << (palperm ? "is" : "isn't") << " a permutation of a palindrome." << std::endl;
  return 0;
}


